lib/rspec/core/mutex.rb

```diff
@@ -0,0 +1,63 @@
+module RSpec
+  module Core
+    # On 1.8.7, it's in the stdlib.
+    # We don't want to load the stdlib, b/c this is a test tool, and can affect the test environment,
+    # causing tests to pass where they should fail.
+    #
+    # So we're transcribing/modifying it from https://github.com/ruby/ruby/blob/v1_8_7_374/lib/thread.rb#L56
+    # Some methods we don't need are deleted.
+    # Anything I don't understand (there's quite a bit, actually) is left in.
+    # Some formating changes are made to appease the robot overlord:
+    #   https://travis-ci.org/rspec/rspec-core/jobs/54410874
+    # @private
+    class Mutex
+      def initialize
+        @waiting = []
+        @locked = false
+        @waiting.taint
+        taint
+      end
+
+      # @private
+      def lock
+        while Thread.critical = true && @locked
+          @waiting.push Thread.current
+          Thread.stop
+        end
+        @locked = true
+        Thread.critical = false
+        self
+      end
+
+      # @private
+      def unlock
+        return unless @locked
+        Thread.critical = true
+        @locked = false
+        begin
+          t = @waiting.shift
+          t.wakeup if t
+        rescue ThreadError
+          retry
+        end
+        Thread.critical = false
+        begin
+          t.run if t
+        rescue ThreadError
+          :noop
+        end
+        self
+      end
+
+      # @private
+      def synchronize
+        lock
+        begin
+          yield
+        ensure
+          unlock
+        end
+      end
+    end unless defined?(::RSpec::Core::Mutex) # Avoid warnings for library wide checks spec
```

Comment:
This is neat! Strange that none of the "yo, I'm done" keywords will get you out of a class binding, otherwise a guard clause would work nicely here.

```ruby
def record_error(code)
  TOPLEVEL_BINDING.eval code
rescue Exception
  return $!
end

record_error 'class A; return; end'  # => #<LocalJumpError: unexpected return>
record_error 'class A; next;   end'  # => #<SyntaxError: <main>: Can't escape from eval with next>
record_error 'class A; break;  end'  # => #<SyntaxError: <main>: Can't escape from eval with break>
```